profile/perf.go

```diff
@@ -0,0 +1,246 @@
+				continue
+			}
+			for _, e := range metricEvents {
+				perfStat.eventSet[e] = 0
+			}
+		}
+	}
+
+	var delimiter, newEventStr string
+	for event := range perfStat.eventSet {
+		newEventStr += delimiter + event
+		delimiter = ","
+	}
+	if newEventStr != "" {
+		perfStat.cmd.Args = append(perfStat.cmd.Args, "-e", newEventStr)
+	}
+
+	perfStat.cmd.Args = append(perfStat.cmd.Args, "--", "sleep", strconv.FormatFloat(executionTime, 'f', -1, 64))
+
+	log.Debugf("Perf command: %s", perfStat.cmd)
+
+	return perfStat
+}
+
+// Run executes perf stat command
+func (p *PerfStat) Run() error {
+	if !isPerfInstalled() {
+		return errors.New("perf is not installed")
+	}
+
+	if p.execTime < 0 {
+		return errors.New("perf execution time is less than 0s")
+	}
+
+	if p.interval < 10 {
+		return errors.New("perf print interval is less than 10ms")
+	}
+
+	if p.interval < 100 {
+		log.Warn("print interval < 100ms. The overhead percentage could be high in some cases. Please proceed with caution.")
+	}
+
+	if err := p.cmd.Start(); err != nil {
+		return err
+	}
+	p.tStart = time.Now()
+
+	return nil
+}
+
+// SetWarmTime sets the time duration until system is warm.
+func (p *PerfStat) SetWarmTime() {
+	p.once.Do(func() {
+		p.warmTime = time.Since(p.tStart).Seconds()
+
+		if p.execTime > 0 && p.warmTime > p.execTime {
+			log.Warn("System warmup time is longer than perf execution time.")
+		}
+	})
+}
+
+// SetTearDownTime sets the time duration until system tears down.
+func (p *PerfStat) SetTearDownTime() {
+	p.tearDownTime = time.Since(p.tStart).Seconds()
+}
+
+// GetResult returns the counters of perf stat
+func (p *PerfStat) GetResult() (map[string]float64, error) {
+	if p.tStart.IsZero() {
+		return nil, errors.New("Perf was not executed, run perf first")
+	}
+
+	// wait for perf command finish
+	timeLeft := (p.execTime - time.Since(p.tStart).Seconds()) * 1e+9
+	time.Sleep(time.Duration(timeLeft))
+
+	log.Debugf("Warm time: %f, Teardown time: %f", p.warmTime, p.tearDownTime)
+	return p.parseResult()
+}
+
+func (p *PerfStat) parseResult() (map[string]float64, error) {
+	data, err := p.readPerfData()
+	if err != nil {
+		return nil, err
+	}
+
+	for _, m := range data {
+		for k, v := range m {
+			p.eventSet[k] += v
+		}
+	}
+
+	for k := range p.eventSet {
+		p.eventSet[k] /= float64(len(data))
+	}
+
+	for _, metric := range p.metrics {
+		events, err := getEvents(metric)
+		if err != nil {
+			return nil, err
+		}
+
+		var params []float64
+		for _, event := range events {
+			params = append(params, p.eventSet[event])
+		}
+
+		result, err := calculateMetric(metric, params...)
+		if err != nil {
+			return nil, err
+		}
+
+		p.eventSet[metric] = result
+	}
+
+	return p.eventSet, nil
+}
+
+func (p *PerfStat) readPerfData() ([]map[string]float64, error) {
+	file, err := os.Open(p.outFile)
+	if err != nil {
+		return nil, errors.New("Perf was failed to execute, check perf events")
+	}
+	defer file.Close()
+
+	scanner := bufio.NewScanner(file)
+	scanner.Scan()
```

Comment:
why duplicate Scans?

Reply:
Because the first line is the title, the second line is empty.